general/restrict-hostnames/rego/src.rego

```diff
@@ -101,8 +97,55 @@ violation[{"msg": msg}] {
 
 		hostpath := is_invalid(host, path)
 	}
-	
+
 	count(invalid_hostpaths) > 0
 
 	msg := sprintf("hostpaths in the virtualservice are not valid for this namespace: %v", [invalid_hostpaths])
 }
+
+# Hostname conflict with other namespace Ingress(es) and hostpath not allowed
+violation[{"msg": msg}] {
+	kind := input.review.kind.kind
+	re_match("^(Ingress|VirtualService)$", kind)
+	re_match("^(networking.k8s.io|networking.istio.io)$", input.review.kind.group)
+
+	hosts := {host | host := input.review.object.spec.rules[_].host} | {host | host := input.review.object.spec.hosts[_]}
+	host := hosts[_]
+	paths := {path | path := input.review.object.spec.rules[_].http.paths[_].path} | ({path | path := input.review.object.spec.http[_].match[_].uri.exact} | {path | path := input.review.object.spec.http[_].match[_].uri.prefix}) | {path | path := input.review.object.spec.http[_].match[_].uri.regex}
+	path := paths[_]
+
+	not is_allowed(host, path)
+
+	ingress_conflicts := {output | conflict := data.inventory.namespace[other_namespace]["networking.k8s.io/v1"]["Ingress"][other_name]
+	conflict.spec.rules[_].host == host
+	conflict.metadata.namespace != input.review.object.metadata.namespace
+	output := concat("/", [other_namespace, other_name])
+	}
+
+	count(ingress_conflicts) > 0
+    msg := sprintf("%v hostname %v conflicts with existing Ingress(es): %v", [kind, host, ingress_conflicts])
+}
+
+
+# Hostname conflict with other namespace VirtualService(s) and hostpath not allowed
+violation[{"msg": msg}] {
+	kind := input.review.kind.kind
+	re_match("^(Ingress|VirtualService)$", kind)
+	re_match("^(networking.k8s.io|networking.istio.io)$", input.review.kind.group)
+
+	hosts := {host | host := input.review.object.spec.rules[_].host} | {host | host := input.review.object.spec.hosts[_]}
+	host := hosts[_]
+	paths := {path | path := input.review.object.spec.rules[_].http.paths[_].path} | ({path | path := input.review.object.spec.http[_].match[_].uri.exact} | {path | path := input.review.object.spec.http[_].match[_].uri.prefix}) | {path | path := input.review.object.spec.http[_].match[_].uri.regex}
+	path := paths[_]
+
+	not is_allowed(host, path)
```

Comment:
For the failure scenarios, I had included [test_ingress_hostname_conflicts](https://github.com/StatCan/gatekeeper-policies/blob/425b6b847e9a191b2903fcd6abb99d01f0ec4d44/general/restrict-hostnames/rego/src_test.rego#L712) and [test_vs_hostname_conflicts](https://github.com/StatCan/gatekeeper-policies/blob/425b6b847e9a191b2903fcd6abb99d01f0ec4d44/general/restrict-hostnames/rego/src_test.rego#L994) which are functioning as expected on my end, reporting a conflict message for an exempted (but not allowed) hostname that exists in another namespace.

Thanks for informing me about `opa fmt`. I'll go ahead and apply it, as well as look into separating the redundant code into a function 😀

Reply:
I had some difficulty separating out the originally quoted code into a function. I attempted the following:
```
conflictable_hosts:= host {
	kind := input.review.kind.kind
	re_match("^(Ingress|VirtualService)$", kind)
	re_match("^(networking.k8s.io|networking.istio.io)$", input.review.kind.group)

	hosts := {host | host := input.review.object.spec.rules[_].host} | {host | host := input.review.object.spec.hosts[_]}
	host := hosts[_]
	paths := ({path | path := input.review.object.spec.rules[_].http.paths[_].path} | ({path | path := input.review.object.spec.http[_].match[_].uri.exact} | {path | path := input.review.object.spec.http[_].match[_].uri.prefix})) | {path | path := input.review.object.spec.http[_].match[_].uri.regex}
	path := paths[_]

	not is_allowed(host, path)
}
```
And referenced it with `host := conflictable_hosts`. While the first test case (conflicting Ingress) would pass, the second (conflicting VirtualService) would error out with `eval_conflict_error: complete rules must not produce multiple outputs`. I gathered that the workaround for this would be to make multiple definitions of `conflictable_hosts`, one checking only an Ingress and another only a VirtualService, but that seemed to add even more bulk... Though maybe something along those lines could be reused in the other violations.

That said while experimenting I found that it is possible to have [one hostname conflict violation instead of two](https://github.com/StatCan/gatekeeper-policies/blob/feat-restrict-ingress-frances/general/restrict-hostnames/rego/src.rego#L105). This is another approach to not duplicating the hostpath code, and results in a single message, for example `{{"msg": "Ingress hostname red.test.com conflicts with existing object(s) in other namespace(s): {\"Ingress/red/red_ingress\", \"VirtualService/red/red_vs\", \"VirtualService/red/red_vs2\"}"}}`.

I'll ponder whether I can make some further improvements.

Reply:
Per discussion in today's team meeting, we can go ahead with this for now, and take a crack at making the rego more elegant later on (perhaps as a Friday activity).